### ISSUE TITLE
MLP design in MaskEstimator

### ISSUE THREAD

**Issue by YoungloLee** [lucidrains/BS-RoFormer#15]
Thanks for your great work.
In the BS-RoFormer paper, the authors mention:
> Each MLP layer consists of a RMS Norm layer, a fully connected layer followed by a Tanh activation, and a fully connected layer followed by a gated linear unit (GLU) layer[29].

However, your MaskEstimator implementation is somewhat different from the paper description.

Here's my implementation (I also checked the number of parameters):
<pre><code>class TanH(Module):
    def __init__(self, dim_in, dim_out):
        super().__init__()
        self.proj = nn.Linear(dim_in, dim_out)

    def forward(self, x):
        x = self.proj(x)
        return x.tanh()</code></pre>
<pre><code>class GLU(Module):
    def __init__(self, dim_in, dim_out):
        super().__init__()
        self.proj = nn.Linear(dim_in, dim_out * 2)

    def forward(self, x):
        x, gate = self.proj(x).chunk(2, dim=-1)
        return x * gate.sigmoid()</code></pre>
<pre><code>class MaskEstimator(Module):
    @beartype
    def __init__(
            self,
            dim,
            dim_inputs: Tuple[int, ...],
            mlp_expansion_factor = 4,
    ):
        super().__init__()
        self.dim_inputs = dim_inputs
        self.to_freqs = ModuleList([])

        for dim_in in dim_inputs:
            net = []
            net.append(TanH(dim, dim * ff_mult))
            net.append(GLU(dim * ff_mult, dim_in))
            self.to_freqs.append(nn.Sequential(*net))

    def forward(self, x):
        x = x.unbind(dim=-2)

        outs = []

        for band_features, to_freq in zip(x, self.to_freqs):
            freq_out = to_freq(band_features)
            outs.append(freq_out)

        return torch.cat(outs, dim=-1)</code></pre>

**Comment by lucidrains** [comment]
@YoungloLee hey Younglo

yes i think you are right

want to check 0.2.6 see if it aligns better? also, do you know if the `tanh` output aligns with the reals, and the `glu` output aligns with the imaginary? i don't have enough domain expertise to know why one mlp output is bounded between `-1` and `1` while the other is not

**Comment by 42speech** [comment]
According to the paper, `tanh` activation output is directly fed into the next fully connected layer with `glu` activation (e.g., MLP: Sequential(RMSNorm, Linear, tanh, Linear, GLU)).
I'm not certain, but it doesn't seem to be related to the real and imaginary components.. It just appears to be the authors' preference for an activation function

**Comment by YoungloLee** [comment]
@lucidrains 

According to the paper, tanh activation output is directly fed into the next fully connected layer with glu activation:

`MLP: Sequential(RMSNorm, Linear, tanh, Linear, GLU))`

I'm not certain, but those two don't seem to be related to the real and imaginary components. It just appears to be the authors' preference for an activation function

**Comment by lucidrains** [comment]
@YoungloLee ahh i totally misunderstood, thank you! i'm used to seeing swish for activation functions

ok, 0.2.8 should be accurate to the paper then!

**Comment by Psarpei** [comment]
Looks pretty well now and the parameters getting really close to the original paper but I think the RMS Norm they mentioned is still missing

**Comment by lucidrains** [comment]
@Psarpei it is taken care of at the end of the attention layers https://github.com/lucidrains/BS-RoFormer/blob/main/bs_roformer/bs_roformer.py#L302

**Comment by lucidrains** [comment]
it is standard to do a final normalization at the end of pre-normalized transformer

**Comment by Psarpei** [comment]
Aaaaah I see thanks :)
